### ISSUE TITLE
iocage clean -a - what does it do?

### ISSUE THREAD

**Issue by xerovis** [iocage/iocage#216]
I can see in the man page that you describe what happens but to someone who isn't an expert at ZFS, it doesn't make any sense. I am running an iocage command and the man page talks about ZFS and basejails. I am just a developer, I have no idea what either of those are.

Developers just want to be able to code, install software and have a container setup and running properly. They don't really have the time to read about how Jails or ZFS works, they want to trust the underlying container software and have their code run. I think if you look at most people who use FreeBSD they want the same. Sure there are the experts who use FreeBSD and they love knowing everything but a large percentage use FreeBSD because "it does things the right way".

**Comment by skarekrow** [comment]
It's best for you to become familiar with the filesystem you use. The man page explains clearly what each flag does and why they are relevant. A dataset is a very common term for ZFS and it is needed to describe what we are doing. This command isn't meant to be ran with reckless abandon, you should know why you're using it. Which means knowing what a basejail is and if you have created any. That last point will be moot soon as we are completely migrating over to them in 2.0 and that will simplify the wording a bit to just "jails".

**Comment by xerovis** [comment]
"It's best for you to become familiar with the filesystem you use."

I see this comment a lot in technical circles and I used to make comments like this myself. I don't actually agree with you, its a good idea for a user of iocage to understand ZFS but it should not be a requirement, which at the moment "seems" to be the case with iocage. Similarly, knowing how to change a tire on your your car is a good idea but its not a requirement of driving a car.

**Comment by dneades** [comment]
@skarekrow When you say that ‘we are completely migrating over to them in 2.0’ (by which I’m presuming that you mean basejails), does that mean that iocage will then cease to support thick jails? Basejails are not presently a good fit for us because they make it difficult to configure a jail fully (e.g. by chrooting into it to run various commands; copying files into it) prior to starting it for the first time (or, indeed, at other times when the jail is not actually running).

(Aside: Having just written the above, it occurs to me that a command that put all the filesystems in place as if the basejail were about to be started, but that did not actually start it, could be generally useful…)

**Comment by skarekrow** [comment]
@dneades Yes that is what that means. We want to simplify everything by only supporting empty and basejails. Can you submit an issue that requests chroot to mount filesystems? That way I don't forget this. I think that will handle your usecase just fine.

**Comment by dneades** [comment]
@skarekrow I have filed #224 to cover `iocage chroot`. While that will solve some of my difficulties with basejails, there are other issues:

1. It will still be difficult to script copying files into (or out of) a basejail’s dataset prior to launching it, because the appropriate paths to use will be dependent upon iocage implementation details. Worse, those paths change if the jail is running, so scripting such things becomes significantly more complicated, as scripts will need to become aware of whether or not the jail is running (with attendant problems if a jail stops or starts while such a script is running).

2. If I am understanding how they work, updates of basejails always require a reboot/restart, even though a reboot is not always required by `freebsd-update`. This is because (again, if I am understanding correctly) updating a basejail involves upgrading the base, then restarting the jail to re-clone its now-upgraded base datasets. Consequently, basejails involve more downtime than thick jails, where a non-reboot-requiring `freebsd-update` can happen within a jail as it would with any other system. (I suppose one could run `freebsd-update` within the jail *as well as* updating the base so that the update will also be present after a reboot, but that seems icky.)

Or, have I missed or misunderstood something?

I like that iocage supports basejails for those who want them, but I suspect I am not the only one who would prefer the conceptual and implementation simplicity of thick jails. The small reduction in disk footprint that basejails provide doesn’t to me seem worth the disadvantages.

**Comment by skarekrow** [comment]
@dneades Thanks for filing that issue! You are overcomplicating things :) We are using overlays for this reason. So to answer these in order:

1. You only need to care about the paths /iocage/jails/UUID/_/* whether a jail is running or not. Which is just as simple as the thick jails were.

2. In order to update the basejail's base, you would need to refetch, you can not run `freebsd-update` inside of a basejail. I suppose a short reboot is more downtime, but we are talking seconds. It's not worth the extra effort to support thick and clone jails with every operation in the future. We decided that it is much easier to support one main jail type (ignoring git jails as they are completely separate for all intents and purposes) when we are making other decisions in the script. It's really just a matter of time, the script is getting much larger and accounting for every gotcha with each jail type is becoming to get laborious, hopefully you understand :+1:

**Comment by dneades** [comment]
@skarekrow Thank you for the clarification – much appreciated!

1. I hadn’t realized you were using `unionfs`, so yes, that does help considerably. The `/iocage/jails/UUID/_/*` layout is new since I last peeked at the source, too! I suspect I’ll be able to wrangle my existing thickjail initialization code to work with this scheme, at least once `iocage chroot` is fixed to mount the filesystems.

2. Yes, the reboot will usually be seconds – assuming, of course, that one doesn’t have processes with long startup times. (That may not be true for some users.) Still, no downtime is even better than a short reboot :-) I do understand the desire to simplify the code, however, and I understand why you’re doing what you are. It’s just that, for my use, if I had to choose between having only thick jails or only basejails, I’d have chosen the former. I understand that other users (e.g. those with very large numbers of jails, where update efficiency starts to matter) may think differently!

**Comment by skarekrow** [comment]
@dneades Yes, I finally decided on a layout :D I also understand, but thankfully it's a minor setback and not something easily worked around. As recloning by it's nature would require us to restart the jail. I think the speed of creation, and upgrading will offset this inconvenience. Sorry your usecase got hit a bit by this :/, not intended of course. We want to see 2.0 lean and mean. Throwing away all the legacy luggage we can. It not only helps us, but contributors as well!

**Comment by dneades** [comment]
:-)
